### ISSUE TITLE
Peft does not really accelerate or improve the training 

### ISSUE THREAD

**Issue by Aktsvigun** [huggingface/peft#925]
### System Info

Python 3.9.0
transformers==4.30.2
peft==0.5.0

CUDA Version: 12.2, NVIDIA A100-SXM4-80GB

### Who can help?

@sgugger

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

Code for running without peft:

```
import os
os.environ["HF_DATASETS_OFFLINE"] = "1"
os.environ["CUDA_VISIBLE_DEVICES"] = "0"

import numpy as np
import pandas as pd
import json

import transformers as ts
from datasets import load_dataset, Dataset, load_metric, load_from_disk, DatasetDict
# from evaluate import load
import torch
from torch import nn


def tokenizing_fn(instance):    
    encoded = tokenizer(instance['sentence'], truncation=True)
    return encoded


def compute_metrics(outputs):
    logits, labels = outputs
    preds = logits.argmax(1)

    metrics_dict = accuracy_metric.compute(
        references=labels, predictions=preds)
    metrics_dict.update(cola_metric.compute(
        references=labels, predictions=preds))
    metrics_dict.update(precision_metric.compute(
        references=labels, predictions=preds))
    metrics_dict.update(recall_metric.compute(
        references=labels, predictions=preds))
    metrics_dict['f1_score'] = 2 * metrics_dict['precision'] * metrics_dict['recall'] /\
    max(metrics_dict['precision'] + metrics_dict['recall'], 1e-6)
    
    return metrics_dict


def model_init():
    return ts.AutoModelForSequenceClassification.from_pretrained(CHECKPOINT, num_labels=2)


CHECKPOINT = 'microsoft/deberta-v3-large'

tokenizer = ts.AutoTokenizer.from_pretrained(CHECKPOINT)

data = load_dataset('glue', 'cola', cache_dir='data')

tokenized_data = data.map(
    tokenizing_fn, batched=True, remove_columns=[x for x in data["train"].column_names if x != 'label']
)
data_collator = ts.DataCollatorWithPadding(tokenizer=tokenizer, padding="longest")

accuracy_metric = load_metric('accuracy', cache_dir='evaluate_metrics')
cola_metric = load_metric('glue', 'cola', cache_dir='evaluate_metrics')
precision_metric = load_metric('precision', cache_dir='evaluate_metrics')
recall_metric = load_metric('recall', cache_dir='evaluate_metrics')

training_args = ts.TrainingArguments(
    output_dir='deberta_output',
    # Batch size args
    num_train_epochs=100,
    per_device_train_batch_size=64,
    per_device_eval_batch_size=64,
    gradient_accumulation_steps=4,
    max_grad_norm=1.,
    # Optimizer args
    learning_rate=2e-5,
    weight_decay=0.01,
    # Scheduler args
    warmup_ratio=0.01,
    # Eval args
    metric_for_best_model='eval_accuracy',
    load_best_model_at_end=True,
    evaluation_strategy='epoch',
    logging_strategy='epoch',
    save_strategy='epoch',
    save_total_limit=10,
    report_to="none",
    fp16=True
)
callbacks = [ts.EarlyStoppingCallback(5)]

trainer = ts.Trainer(
    model_init=model_init,
    args=training_args,
    data_collator=data_collator,
    train_dataset=tokenized_data["train"],
    eval_dataset=tokenized_data["validation"],
    callbacks=callbacks,
    compute_metrics=compute_metrics
)

trainer.train()

test_results = trainer.evaluate(tokenized_data["test"])
print(test_results)

import json
with open("test_results_fce.json", "w") as f:
    json.dump(test_results, f)
```

Code for running with peft:

```
import os
os.environ["HF_DATASETS_OFFLINE"] = "1"
os.environ["CUDA_VISIBLE_DEVICES"] = "0"

import numpy as np
import pandas as pd
import json

import transformers as ts
from datasets import load_dataset, Dataset, load_metric, load_from_disk, DatasetDict
import torch
from torch import nn

from peft import (
    LoraConfig,
    get_peft_model,
    prepare_model_for_kbit_training,
    AutoPeftModelForSequenceClassification,
)


def tokenizing_fn(instance):    
    encoded = tokenizer(instance['sentence'], truncation=True)
    return encoded


def compute_metrics(outputs):
    logits, labels = outputs
    preds = logits.argmax(1)

    metrics_dict = accuracy_metric.compute(
        references=labels, predictions=preds)
    metrics_dict.update(cola_metric.compute(
        references=labels, predictions=preds))
    metrics_dict.update(precision_metric.compute(
        references=labels, predictions=preds))
    metrics_dict.update(recall_metric.compute(
        references=labels, predictions=preds))
    metrics_dict['f1_score'] = 2 * metrics_dict['precision'] * metrics_dict['recall'] /\
    max(metrics_dict['precision'] + metrics_dict['recall'], 1e-6)
    
    return metrics_dict

def create_peft_config():
    """
    Create Parameter-Efficient Fine-Tuning config for your model
    :param modules: Names of the modules to apply Lora to
    """
    config = LoraConfig(
        task_type="SEQ_CLS",
        inference_mode=False,
        r=8,
        lora_alpha=16,
        lora_dropout=0.1
    )

    return config

CHECKPOINT = 'microsoft/deberta-v3-large'

peft_config = create_peft_config()
model = ts.AutoModelForSequenceClassification.from_pretrained(CHECKPOINT, num_labels=2)
model = get_peft_model(model, peft_config)
model.print_trainable_parameters()

tokenizer = ts.AutoTokenizer.from_pretrained(CHECKPOINT)

data = load_dataset('glue', 'cola', cache_dir='data')

tokenized_data = data.map(
    tokenizing_fn, batched=True, remove_columns=[x for x in data["train"].column_names if x != 'label']
)
data_collator = ts.DataCollatorWithPadding(tokenizer=tokenizer, padding="longest")

accuracy_metric = load_metric('accuracy', cache_dir='evaluate_metrics')
cola_metric = load_metric('glue', 'cola', cache_dir='evaluate_metrics')
precision_metric = load_metric('precision', cache_dir='evaluate_metrics')
recall_metric = load_metric('recall', cache_dir='evaluate_metrics')

training_args = ts.TrainingArguments(
    output_dir='deberta_output',
    # Batch size args
    num_train_epochs=100,
    per_device_train_batch_size=64,
    per_device_eval_batch_size=64,
    gradient_accumulation_steps=4,
    max_grad_norm=1.,
    # Optimizer args
    learning_rate=2e-5,
    weight_decay=0.01,
    # Scheduler args
    warmup_ratio=0.01,
    # Eval args
    metric_for_best_model='eval_accuracy',
    load_best_model_at_end=True,
    evaluation_strategy='epoch',
    logging_strategy='epoch',
    save_strategy='epoch',
    save_total_limit=10,
    report_to="none",
    fp16=True
)
callbacks = [ts.EarlyStoppingCallback(5)]

trainer = ts.Trainer(
    model_init=model_init,
    args=training_args,
    data_collator=data_collator,
    train_dataset=tokenized_data["train"],
    eval_dataset=tokenized_data["validation"],
    callbacks=callbacks,
    compute_metrics=compute_metrics
)

trainer.train()

test_results = trainer.evaluate(tokenized_data["test"])
print(test_results)

import json
with open("test_results_fce.json", "w") as f:
    json.dump(test_results, f)
```

**Tl;dr:** For training with peft, I only substituted the part 
```
def model_init():
    return ts.AutoModelForSequenceClassification.from_pretrained(CHECKPOINT, num_labels=2)
```

with the part
```
def create_peft_config():
    """
    Create Parameter-Efficient Fine-Tuning config for your model
    :param modules: Names of the modules to apply Lora to
    """
    config = LoraConfig(
        task_type="SEQ_CLS",
        inference_mode=False,
        r=8,
        lora_alpha=16,
        lora_dropout=0.1
    )

    return config

peft_config = create_peft_config()
model = ts.AutoModelForSequenceClassification.from_pretrained(CHECKPOINT, num_labels=2)
model = get_peft_model(model, peft_config)
model.print_trainable_parameters()
```
and made the corresponding change in the `Trainer` (model object instead of the function).

### Expected behavior

I am running the DeBERTa model on the CoLA dataset and wanted to use peft to accelerate the training.
I used the LoRA method as described in the [official example](https://github.com/huggingface/peft/blob/main/examples/sequence_classification/Prompt_Tuning.ipynb) and therefore trained only 0.18% of parameters as stated by `model.` However, the estimated time after 2 iterations is **29 minutes without peft** (conventional fine-tuning) against **32 minutes with peft**. In addition, the model fine-tuned with peft is of significantly worse quality. While I clearly understand that the model quality may be a subject to the dataset specific, I expect `peft` and the implemented methods acсelerate the training instead of making it even longer. Therefore, I have a question: (1) do I have any errors (maybe I did not use the full power of `peft`) or (2) can there potentially be bugs inside `peft` implementation of LoRA or (3) is it a subject to the LoRA method itself?
Thanks a lot in advance!

**Comment by younesbelkada** [comment]
Hi @Aktsvigun 
Thanks for the issue, note that PEFT methods are not really meant to accelerate training for small models. Note that for models such as DeBERTa or RoBERTa, this is expected as the overhead introduced in LoRA operation becomes more important with respect to other operations.
PEFT and more specifically LoRA, IA3 and AdaLoRA becomes extremely interesting and useful for large-scale models (>1b parameters) because it enables training these models at low cost, for example on a Google colab it is possible to fine-tune up to 20B scale models (!) if you use 4bit quantization.
Let us know if anything is unclear

**Comment by Aktsvigun** [comment]
@younesbelkada thank you for the reply! May I kindly ask you what overhead you are speaking about?

**Comment by younesbelkada** [comment]
Recall the LoRA computation below

![lora-animated](https://github.com/huggingface/peft/assets/49240599/dd21250d-4474-49ea-a054-f8e55fd4fb38)

What I meant by the overhead is the computation on the right, if the hidden dimension is quite low, which is the case for small models as DeBERTa and RoBERTa that overhead will be quite considerable compared to the frozen matmul on the left

**Comment by Aktsvigun** [comment]
Thanks a lot for your reply, Younes!
